### PR TITLE
Abstract logic for collecting model config files + Begin ignoring yaml files that are .gitignore'd

### PR DESCRIPTION
**Pull request by QMalcolm**
Other projects use MetricFlow. Some of these have custom yaml config filtering logic. By abstracting out the logic for collecting model config files we:
- Make it reusable throughout the code base
- Allow people to use it and do additional filter on the results
Also, by having functions for linting model config files paths and functions for building a model from config file paths, we allow people more flexibility than by only building the model from a directory.

Additionally, in this PR we begin using GItPython to check if a config dir is a repository, and if so ignore YAML files which are ignored via the repository's `.gitignore` file. This was an issue which cropped up because sometimes dependencies of projects include YAML files in their package.

### CODE REVIEW

**Review by tlento — APPROVED** [review]
Thank you for making this so easy to review! I look forward to the day when we can delete the old functions. Red diffs bring me joy.

File: metricflow/model/parsing/dir_to_model.py [hunk 1]
```diff
@@ -87,11 +82,31 @@ def parse_directory_of_yaml_files_to_model(
...
+
+
+def parse_directory_of_yaml_files_to_model(
+    directory: str,
+    template_mapping: Optional[Dict[str, str]] = None,
+    apply_pre_transformations: Optional[bool] = True,
+    apply_post_transformations: Optional[bool] = True,
+    raise_issues_as_exceptions: bool = True,
+) -> ModelBuildResult:
+    """Parse files in the given directory to a TMdoModel.
```
[the remark this answers is not in the document]
Comment (tlento) [reply metricflow/model/parsing/dir_to_model.py:107]: Oh, forgot to ask, can you change that to UserConfiguredModel or whatever we actually call it these days? Thanks!

**Comment by QMalcolm** [comment]
> Thank you for making this so easy to review! I look forward to the day when we can delete the old functions. Red diffs bring me joy.

I think the plan would be
1) Mark the functions as deprecated
2) Wait until we're comfortable to do a version bump denoting a breaking change

**Comment by QMalcolm** [comment]
Added a commit ([abc3bf8](https://github.com/transform-data/metricflow/pull/215/commits/abc3bf8da41afe93f1ac12e496bc34fcd04f0869)) for fixing the TMdoModel referenced in the existing documentation (new commit because I don't think it wholly belonged to any other commit, more of a tangential fix). Additionally, fixup'd a extra docstring details into [c9814a8](https://github.com/transform-data/metricflow/pull/215/commits/c9814a8a12f0ae10f420bcb86c112cf37ade476d)

**Pull request merged**